### PR TITLE
fix: adjust the implementation to the new spec version

### PR DESCRIPTION
**Pull request by Iloniavo**

### CODE REVIEW

File: src/main/java/school/hei/haapi/endpoint/rest/controller/FeeController.java [hunk 1]
```diff
@@ -76,4 +82,25 @@ public List<Fee> getFees(
...
+  public List<FeeTemplate> getPredefinedFeeTypes(
+          @RequestParam(value = "name", required = false) String name,
+          @RequestParam(value = "amount", required = false) Integer amount,
+          @RequestParam(value = "number_of_payments", required = false) Integer numberOfPayments,
+          @RequestParam(value = "page", defaultValue = "1") PageFromOne page,
+          @RequestParam(value = "page_size", defaultValue = "10") BoundedPageSize pageSize) {
+    return feeTemplateService.getFeeTypes(name, amount, numberOfPayments, page, pageSize).stream()
+            .map(feeTemplateMapper::toRest)
+            .collect(toUnmodifiableList());
+  }
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/endpoint/rest/controller/FeeController.java]: getFeeTemplates

File: src/main/java/school/hei/haapi/endpoint/rest/controller/FeeController.java [hunk 2]
```diff
@@ -76,4 +82,25 @@ public List<Fee> getFees(
         .map(feeMapper::toRestFee)
         .collect(toUnmodifiableList());
   }
+
+  @GetMapping("/fees/templates")
+  public List<FeeTemplate> getPredefinedFeeTypes(
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/endpoint/rest/controller/FeeController.java]: getFeeTemplates

File: src/main/java/school/hei/haapi/endpoint/rest/controller/FeeController.java [hunk 3]
```diff
@@ -76,4 +82,25 @@ public List<Fee> getFees(
...
+          @RequestParam(value = "number_of_payments", required = false) Integer numberOfPayments,
+          @RequestParam(value = "page", defaultValue = "1") PageFromOne page,
+          @RequestParam(value = "page_size", defaultValue = "10") BoundedPageSize pageSize) {
+    return feeTemplateService.getFeeTypes(name, amount, numberOfPayments, page, pageSize).stream()
+            .map(feeTemplateMapper::toRest)
+            .collect(toUnmodifiableList());
+  }
+
+  @PutMapping("/fees/templates/{id}")
+  public FeeTemplate createOrUpdatePredefinedFeetype(
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/endpoint/rest/controller/FeeController.java]: createOrUpdateFeeTemplates

File: src/main/java/school/hei/haapi/endpoint/rest/controller/FeeController.java [hunk 4]
```diff
@@ -76,4 +82,25 @@ public List<Fee> getFees(
...
+          @PathVariable String id, @RequestBody CrupdateFeeTemplate feeType) {
+    return feeTemplateMapper.toRest(
+            feeTemplateService.createOrUpdateFeeTypes(feeTemplateMapper.toDomain(feeType)));
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/endpoint/rest/controller/FeeController.java]: createOrUpdateFeeTemplates

File: src/main/java/school/hei/haapi/endpoint/rest/mapper/FeeTemplateTypeEnumMapper.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+import school.hei.haapi.endpoint.rest.model.FeeTemplate;
+import school.hei.haapi.model.exception.BadRequestException;
+
+@Component
+public class FeeTemplateTypeEnumMapper {
+
+  public CrupdateFeeTemplate.TypeEnum toRestEnum(FeeTemplate.TypeEnum domain) {
+    if (domain == null) {
+      return null;
+    }
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/endpoint/rest/mapper/FeeTemplateTypeEnumMapper.java]: no

File: src/main/java/school/hei/haapi/endpoint/rest/mapper/FeeTemplateTypeEnumMapper.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+      case STUDENT_INSURANCE -> CrupdateFeeTemplate.TypeEnum.STUDENT_INSURANCE;
+      case REMEDIAL_COSTS -> CrupdateFeeTemplate.TypeEnum.REMEDIAL_COSTS;
+      default -> throw new BadRequestException("Unexpected type " + domain);
+    };
+  }
+
+  public FeeTemplate.TypeEnum toDomainEnum(CrupdateFeeTemplate.TypeEnum rest) {
+    if (rest == null) {
+      return null;
+    }
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/endpoint/rest/mapper/FeeTemplateTypeEnumMapper.java]: no

File: doc/api.yml
Comment (Mahefaa) [inline doc/api.yml:1]: add getById also

**Review by Mahefaa — CHANGES_REQUESTED** [review]

File: doc/api.yml
Comment (Iloniavo) [reply doc/api.yml:1]: Got all of it

File: src/main/java/school/hei/haapi/model/FeeTemplate.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+  private String name;
+
+  @CreationTimestamp private Instant creationDatetime;
+
+  private Integer amount;
+
+  private Integer numberOfPayments;
+
+  @Type(type = "pgsql_enum")
+  @Enumerated(EnumType.STRING)
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/model/FeeTemplate.java]: static import

File: src/test/java/school/hei/haapi/integration/FeeTemplateIT.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
+package school.hei.haapi.integration;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.springframework.boot.test.context.SpringBootTest.WebEnvironment.RANDOM_PORT;
+import static school.hei.haapi.integration.conf.TestUtils.*;
```
Comment (Mahefaa) [inline src/test/java/school/hei/haapi/integration/FeeTemplateIT.java]: rm import *

File: src/test/java/school/hei/haapi/integration/conf/TestUtils.java [hunk 9]
```diff
@@ -599,6 +603,34 @@ public static StudentGrade studentGrade7() {
         .grade(grade7());
   }
 
+  public static FeeTemplate feeTemplate1() {
+    return new FeeTemplate()
+        .id(FEE_TEMPLATE1_ID)
+        .name("annuel x9")
+        .numberOfPayments(9)
+        .amount(200000)
+        .type(FeeTemplate.TypeEnum.TUITION)
```
Comment (Mahefaa) [inline src/test/java/school/hei/haapi/integration/conf/TestUtils.java]: static import

File: src/test/java/school/hei/haapi/integration/conf/TestUtils.java [hunk 10]
```diff
@@ -599,6 +603,34 @@ public static StudentGrade studentGrade7() {
...
+        .creationDatetime(Instant.parse("2022-11-08T08:25:24.00Z"));
+  }
+
+  public static CrupdateFeeTemplate createFeeTemplate2() {
+    return new CrupdateFeeTemplate()
+        .id(FEE_TEMPLATE2_ID)
+        .name("annuel x1")
+        .numberOfPayments(1)
+        .amount(10000)
+        .type(CrupdateFeeTemplate.TypeEnum.TUITION);
```
Comment (Mahefaa) [inline src/test/java/school/hei/haapi/integration/conf/TestUtils.java]: static import

File: src/test/java/school/hei/haapi/integration/conf/TestUtils.java [hunk 11]
```diff
@@ -599,6 +603,34 @@ public static StudentGrade studentGrade7() {
...
+  }
+
+  public static FeeTemplate feeTemplate2() {
+    return new FeeTemplate()
+        .id(FEE_TEMPLATE2_ID)
+        .name("annuel x1")
+        .numberOfPayments(1)
+        .amount(10000)
+        .type(FeeTemplate.TypeEnum.TUITION);
```
Comment (Mahefaa) [inline src/test/java/school/hei/haapi/integration/conf/TestUtils.java]: static import

File: src/main/java/school/hei/haapi/service/FeeTemplateService.java [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+      BoundedPageSize pageSize) {
+    Pageable pageable =
+        PageRequest.of(page.getValue() - 1, pageSize.getValue(), Sort.by(DESC, "creationDatetime"));
+    return feeTemplateDao.findByCriteria(name, totalAmount, numberOfMonths, pageable);
+  }
+
+  public FeeTemplate getFeeTemplateById(String id) {
+    return feeTemplateRepository.findById(id).orElseThrow(
+            () -> {
+              throw new NotFoundException("{\"type\":\"404 NOT_FOUND\",\"message\":\"FeeTemplate with id: \" + id_not_existing + \" not found\"}");
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/service/FeeTemplateService.java]: () -> new NotFoundException(String);

**Review by Mahefaa — CHANGES_REQUESTED** [review]

File: doc/api.yml [hunk 13]
```diff
@@ -2449,6 +2477,13 @@ components:
         creation_datetime:
           type: string
           format: date-time
+    TypeEnum:
+      type: string
+      enum:
+        - TUITION
+        - HARDWARE
+        - STUDENT_INSURANCE
+        - REMEDIAL_COSTS
```
Comment (Mahefaa) [inline doc/api.yml]: FeeTypeEnum

File: src/main/java/school/hei/haapi/endpoint/rest/mapper/FeeTemplateTypeEnumMapper.java [hunk 14]
```diff
@@ -0,0 +1,12 @@
...
+import org.springframework.stereotype.Component;
+import school.hei.haapi.endpoint.rest.model.CrupdateFeeTemplate;
+import school.hei.haapi.endpoint.rest.model.FeeTemplate;
+import school.hei.haapi.endpoint.rest.model.TypeEnum;
+import school.hei.haapi.model.exception.BadRequestException;
+
+@Component
+public class FeeTemplateTypeEnumMapper {
+
+}
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/endpoint/rest/mapper/FeeTemplateTypeEnumMapper.java]: what ?

File: doc/api.yml [hunk 15]
```diff
@@ -1566,24 +1566,24 @@ paths:
         '500':
           $ref: '#/components/responses/500'
 
-  /fees/types:
+  /fees/templates:
     get:
       tags:
         - Paying
-      summary: Get all predefined types of fee
+      summary: Get all templates of fee
```
Comment (Mahefaa) [inline doc/api.yml]: get all fee templates

File: src/main/java/school/hei/haapi/model/FeeTemplate.java [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/model/FeeTemplate.java]: static import !

**Review by Mahefaa — CHANGES_REQUESTED** [review]

File: doc/api.yml [hunk 17]
```diff
@@ -1624,20 +1653,20 @@ paths:
           schema:
             type: string
       description: Update fee type when `id` exist, create it otherwise.
```
Comment (Mahefaa) [inline doc/api.yml]: update IF exists not WHEN exists

File: src/main/java/school/hei/haapi/model/Fee.java [hunk 18]
```diff
@@ -48,12 +50,12 @@ public class Fee implements Serializable {
   private User student;
 
   @Type(type = "pgsql_enum")
-  @Enumerated(EnumType.STRING)
+  @Enumerated(STRING)
   private school.hei.haapi.endpoint.rest.model.Fee.StatusEnum status;
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/model/Fee.java]: try to use StatusEnum directly also here without giving the full package name

**Review by Mahefaa — CHANGES_REQUESTED** [review]
few changes left x) good job :))

File: src/main/java/school/hei/haapi/endpoint/rest/mapper/FeeMapper.java [hunk 19]
```diff
@@ -87,12 +90,12 @@ public List<school.hei.haapi.model.Fee> toDomainFee(User student, List<CreateFee
...
 
-  private Fee.TypeEnum toDomainFeeType(CreateFee.TypeEnum createFeeType) {
+  private FeeTypeEnum toDomainFeeType(FeeTypeEnum createFeeType) {
     switch (createFeeType) {
       case TUITION:
-        return Fee.TypeEnum.TUITION;
+        return TUITION;
       case HARDWARE:
-        return Fee.TypeEnum.HARDWARE;
+        return HARDWARE;
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/endpoint/rest/mapper/FeeMapper.java:99]: where are the other two ?

File: src/main/java/school/hei/haapi/repository/FeeTemplateRepository.java [hunk 20]
```diff
@@ -0,0 +1,12 @@
...
+
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+import school.hei.haapi.model.FeeTemplate;
+
+import java.util.Optional;
+
+@Repository
+public interface FeeTemplateRepository extends JpaRepository<FeeTemplate, String> {
+    Optional<FeeTemplate> findById(String s);
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/repository/FeeTemplateRepository.java]: is it really needed ??

**Review by Mahefaa — CHANGES_REQUESTED** [review]

File: src/main/java/school/hei/haapi/service/FeeTemplateService.java [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+        PageRequest.of(page.getValue() - 1, pageSize.getValue(), Sort.by(DESC, "creationDatetime"));
+    return feeTemplateDao.findByCriteria(name, totalAmount, numberOfMonths, pageable);
+  }
+
+  public FeeTemplate getFeeTemplateById(String id) {
+    return feeTemplateRepository.findById(id).orElseThrow(
+            () -> {
+              throw new NotFoundException("FeeTemplate with id: " + id + " not found");
+            });
+  }
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/service/FeeTemplateService.java]: orElseThrow( () -> new NotFoundExcepton("message"));

File: src/test/java/school/hei/haapi/integration/FeeTemplateIT.java (hunk 8 above)
Comment (Mahefaa) [reply src/test/java/school/hei/haapi/integration/FeeTemplateIT.java]: do not use import * !!

File: src/test/java/school/hei/haapi/integration/FeeTemplateIT.java [hunk 22]
```diff
@@ -0,0 +1,83 @@
...
+    setUpCognito(cognitoComponentMock);
+  }
+
+  @Test
+  void manager_get_fee_templates() throws ApiException {
+    ApiClient managerClient = anApiClient(MANAGER1_TOKEN);
+    PayingApi api = new PayingApi(managerClient);
+    List<FeeTemplate> actual = api.getFeeTemplates(null, 200000, 9, 1, 10);
+    assertEquals(1, actual.size());
+    assertTrue(actual.contains(feeTemplate1()));
```
Comment (Mahefaa) [inline src/test/java/school/hei/haapi/integration/FeeTemplateIT.java]: separate given when then sides with blank line

File: src/test/java/school/hei/haapi/integration/FeeTemplateIT.java [hunk 23]
```diff
@@ -0,0 +1,83 @@
...
+
+  @Test
+  void manager_create_fee_templates() throws ApiException {
+    ApiClient managerClient = anApiClient(MANAGER1_TOKEN);
+    PayingApi api = new PayingApi(managerClient);
+    FeeTemplate createdFeeTemplate =
+        api.crupdateFeeTemplate(FEE_TEMPLATE2_ID, createFeeTemplate2());
+    assertEquals(feeTemplate2().getName(), createdFeeTemplate.getName());
+    assertEquals(feeTemplate2().getAmount(), createdFeeTemplate.getAmount());
+    assertEquals(feeTemplate2().getNumberOfPayments(), createdFeeTemplate.getNumberOfPayments());
```
Comment (Mahefaa) [inline src/test/java/school/hei/haapi/integration/FeeTemplateIT.java]: separate given when then sides with blank line

File: src/test/java/school/hei/haapi/integration/FeeTemplateIT.java [hunk 24]
```diff
@@ -0,0 +1,83 @@
...
+  }
+
+  @Test
+  void get_fee_template_by_id_existing() throws ApiException{
+    ApiClient managerClient = anApiClient(MANAGER1_TOKEN);
+    PayingApi api = new PayingApi(managerClient);
+    FeeTemplate actual = api.getFeeTemplateById(FEE_TEMPLATE1_ID);
+    assertEquals(feeTemplate1(), actual);
```
Comment (Mahefaa) [inline src/test/java/school/hei/haapi/integration/FeeTemplateIT.java]: separate given when then sides with blank line

File: src/test/java/school/hei/haapi/integration/conf/TestUtils.java [hunk 25]
```diff
@@ -599,6 +604,34 @@ public static StudentGrade studentGrade7() {
...
+
+  public static FeeTemplate feeTemplate2() {
+    return new FeeTemplate()
+        .id(FEE_TEMPLATE2_ID)
+        .name("annuel x1")
+        .numberOfPayments(1)
+        .amount(10000)
+        .type(TUITION);
+  }
+
```
Comment (Mahefaa) [inline src/test/java/school/hei/haapi/integration/conf/TestUtils.java]: test the other types also

**Review by Mahefaa — CHANGES_REQUESTED** [review]

File: src/main/java/school/hei/haapi/endpoint/rest/mapper/FeeTemplateMapper.java [hunk 26]
```diff
@@ -0,0 +1,31 @@
...
+  public school.hei.haapi.endpoint.rest.model.FeeTemplate toRest(FeeTemplate domain) {
+    return new school.hei.haapi.endpoint.rest.model.FeeTemplate()
+        .id(domain.getId())
+        .type(domain.getType())
+        .amount(domain.getAmount())
+        .name(domain.getName())
+        .creationDatetime(domain.getCreationDatetime())
+        .numberOfPayments(domain.getNumberOfPayments());
+  }
+}
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/endpoint/rest/mapper/FeeTemplateMapper.java]: import the Domain with package full name but the rest with only the className

**Review by Mahefaa — CHANGES_REQUESTED** [review]

**Review by Mahefaa — APPROVED** [review]

**Pull request merged**